### DISCUSSION TITLE
Incorrect RigidBody rotation angle around the axes (Euler angles)

### DISCUSSION THREAD

**Discussion by Apsapeh** [DanielChappuis/reactphysics3d#358] [Q&A] [answered]
Hello, I'm writing a retro game engine, I chose your physics engine as it is quite lightweight and productive for old systems. But when testing a simple scene with a floor and a cube falling on it rotated around its axis, the falling cube behaves very strangely. I attach a video, it shows the strange behaviour of the cube, when the cube rises, I give it acceleration on the Y axis. In the terminal, the video shows the cube rotation angle and rotation coefficient on one line, and the cube position on the Y axis on the other line. Sorry for bad English, I use a translator.

https://github.com/DanielChappuis/reactphysics3d/assets/89659777/19aa09e5-2ed0-46b2-bb2d-d0203034cc95

### Code for the creation of physical bodies
```cpp
PhysicsCommon physicsCommon;
PhysicsWorld* physicsWorld = physicsCommon.createPhysicsWorld();

Vector3 floor_body_vec(0, 0, 0);
Quaternion floor_body_quat = Quaternion::fromEulerAngles(0, 0, 0);
Transform floor_body_trans(floor_body_vec, floor_body_quat);
RigidBody* floor_body = physicsWorld->createRigidBody(floor_body_trans);

BoxShape* floor_col_shape = physicsCommon.createBoxShape({50.0f, 1.0f, 50.0f});
Collider* floor_col = floor_body->addCollider(floor_col_shape, Transform::identity());


Vector3 rbody_body_vec(0, 6, 0);
Quaternion rbody_body_quat = Quaternion::fromEulerAngles(1, 0, 3.9);
Transform rbody_body_trans(rbody_body_vec, rbody_body_quat);
RigidBody* rbody_body = physicsWorld->createRigidBody(rbody_body_trans);

BoxShape* rbody_col_shape = physicsCommon.createBoxShape({1.0f, 1.0f, 1.0f});
Collider* rbody_col = rbody_body->addCollider(rbody_col_shape, Transform::identity());

Vector3 rbodyPosVect = floor_body->getTransform().getPosition();
```

### Code for outputting the angle and rotation of the cube mesh
```cpp
physicsWorld->update((float)delta);

decimal dec;
Vector3 vec;
rbody_body->getTransform().getOrientation().getRotationAngleAxis(dec, vec);
std::cout << dec * vec.x * 180.0 / M_PI << " | " << vec.y * dec * 180 / M_PI << " | " << vec.z * dec * 180 / M_PI<< " | " << dec << "\n";

Vector3 rbodyPosVect = rbody_body->getTransform().getPosition();
std::cout << "Rbody Y pos: " << rbodyPosVect.y << "\n";
mesh1.setPosition(rbodyPosVect.x, rbodyPosVect.y, rbodyPosVect.z);
mesh1.setRotation(vec.x * dec * 180.0f / M_PI,
                      vec.y * dec * 180.0f / M_PI,
                      vec.z * dec * 180.0f / M_PI);
```

**Comment by DanielChappuis** [comment]
First, you should try to compute a correct inertia tensor for your body (as described [here](https://www.reactphysics3d.com/usermanual.html#x1-4300010.8.3)) using the following method after you have added the collider.

`RigidBody::updateMassPropertiesFromColliders()`

Secondly, I am not sure I understand. Are you manually rotating/accelerating the body or is it simply falling down on the floor?

**Reply by Apsapeh** [reply]
Thanks, I'll try the first item when I get home. 
On the second point - I initially set the angle of rotation around the X and Z axes for the falling cube via Quaternion::fromEulerAngles(), and then it falls under the action of gravity. At most, what I do when it is already created is to give it acceleration along the y-axis.
```cpp
if (engine->Input.isKeyPressed(GLFW_KEY_LEFT_SHIFT))
        rbody_body->setLinearVelocity({0, 10, 0});
```

**Reply by Apsapeh** [reply]
Sorry I haven't replied in a while, been busy studying at university. I did what you said, but only with automatic calculation. The cube rotated a little faster, but the problem persisted. I did more tests and if you do not rotate the cube mesh, but just change the position in accordance with the coordinates of the physical body of the cube, everything works correctly. Maybe I'm not getting the rotation of the physical body of the cube in degrees correctly?  I have attached a video with and without rotation.

I should also mention that I don't use the built library via `ccmake + make`, but build from source code using the `xmake` build system.
 
[Here is a link to the source code](https://github.com/Apsapeh/Old3DEngine/blob/master/examples/engine_test_main.cpp) of the program where I use your library for physics simulation and my 3D engine for visualization. The `_physics_process` function is called 60 +- 0.2 times per second. If you have free time and want to build a project, you only need `xmake` installed from dependencies (if you have Linux or MacOS, it's more difficult with Windows). To build, write `xmake` in the root directory of the repository and `xmake run -w .` to run it.

## With the rotation of the mesh:

https://github.com/DanielChappuis/reactphysics3d/assets/89659777/b5667cf0-b977-4c90-b51f-2e48359ece64

## Without rotating the mesh

https://github.com/DanielChappuis/reactphysics3d/assets/89659777/856c92ec-cbdc-492b-92b5-40311e26fe6c

**Comment by DanielChappuis** [comment]
Sorry but I really have a hard time understanding what is going on in your simulation and what you are trying to do. 
Are you constantly rotating the cube while it falls or are you simply rotating it at the beginning before it falls and then don't rotate it anymore?

**Comment by Apsapeh** [comment]
I rotate it in the beginning before it drops, and then I have no effect on it at all.

I did some additional testing and noticed that `Quaternion::getRotationAngleAxis()` returns strange values. 
##### Example: 
```cpp
Quaternion quaternion = Quaternion::fromEulerAngles(1, 0, 1);
decimal angle;
Vector3 rot_axis;
quaternion.getRotationAngleAxis(angle, rot_axis);
std::cout << rot_axis.x * angle << " | "
          << rot_axis.y * angle << " | "
          << rot_axis.z * angle << "   Angle: "
          << angle << "\n";
```
##### Console output:
`0.912519 | 0.498512 | 0.912519   Angle: 1.38344`

**Comment by DanielChappuis** [comment, marked as answer]
This method return the axis of rotation and angle of rotation (in radians) of a quaternion. What is strange about those values?

**Comment by Apsapeh** [comment]
Ah, now I get it, I thought this method was returning Euler angles. Sorry for wasting your time, I'll try to convert quaterion to Euler angles or use quaterion at all.
